src/parser/symtab.ts

```diff
@@ -0,0 +1,1738 @@
+        }
+        return this._funcParams;
+    }
+
+    get_locals(): string[] {
+        assert(this.get_type() == "function", "get_locals only valid for function scopes");
+        if (!this._funcLocals) {
+            this._funcLocals = this._identsMatching((x) => !!(x & SYMTAB_CONSTS.DEF_BOUND));
+        }
+        return this._funcLocals;
+    }
+
+    get_globals(): string[] {
+        assert(this.get_type() == "function", "get_globals only valid for function scopes");
+        if (!this._funcGlobals) {
+            this._funcGlobals = this._identsMatching(function (x) {
+                var masked = (x >> SYMTAB_CONSTS.SCOPE_OFFSET) & SYMTAB_CONSTS.SCOPE_MASK;
+                return masked == SYMTAB_CONSTS.GLOBAL_IMPLICIT || masked == SYMTAB_CONSTS.GLOBAL_EXPLICIT;
+            });
+        }
+        return this._funcGlobals;
+    }
+
+    get_frees() {
+        assert(this.get_type() == "function", "get_frees only valid for function scopes");
+        if (!this._funcFrees) {
+            this._funcFrees = this._identsMatching(function (x) {
+                var masked = (x >> SYMTAB_CONSTS.SCOPE_OFFSET) & SYMTAB_CONSTS.SCOPE_MASK;
+                return masked == SYMTAB_CONSTS.FREE;
+            });
+        }
+        return this._funcFrees;
+    }
+
+    get_methods() {
+        var i;
+        var all;
+        assert(this.get_type() == "class", "get_methods only valid for class scopes");
+        if (!this._classMethods) {
+            // todo; uniq?
+            all = [];
+            for (i = 0; i < this.children.length; ++i) {
+                all.push(this.children[i].name);
+            }
+            all.sort();
+            this._classMethods = all;
+        }
+        return this._classMethods;
+    }
+
+    get_scope(name: string): number {
+        //print("getScope");
+        //for (var k in this.symFlags) print(k);
+        var v = this.symbols[name];
+        if (v === undefined) {
+            return 0;
+        }
+        return (v >> SYMTAB_CONSTS.SCOPE_OFFSET) & SYMTAB_CONSTS.SCOPE_MASK;
+    }
+}
+
+function _Py_Mangle(privateobj: string | null, ident: string) {
+    /* Name mangling: __private becomes _classname__private.
+       This is independent from how the name is used. */
+    if (privateobj === null /* !PyUnicode_Check(privateobj) || */ || !ident.startsWith("__")) {
+        return ident;
+    }
+    /* Don't mangle __id__ or names with dots.
+          The only time a name with a dot can occur is when
+          we are compiling an import statement that has a
+          package name.
+          TODO(jhylton): Decide whether we want to support
+          mangling of the module name, e.g. __M.X.
+       */
+    if (ident.endsWith("__") || ident.indexOf(".") != -1) {
+        return ident; /* Don't mangle __whatever__ */
+    }
+
+    /* Strip leading underscores from class name */
+    const stripped = privateobj.replace(/^_+/, "");
```

Comment:
Any regex we should create in the module rather than the function for performance